qdax/core/emitters/pbt_me_emitter.py

```diff
@@ -0,0 +1,408 @@
+        best_genotypes = jax.tree_util.tree_map(
+            lambda x: jax.random.choice(
+                sub_key, x, shape=(len(fitnesses),), replace=True
+            ),
+            best_genotypes_local,
+        )
+        best_training_states, best_replay_buffers = best_genotypes
+
+        # Resample hyper-params
+        best_training_states = jax.vmap(
+            best_training_states.__class__.resample_hyperparams
+        )(best_training_states)
+
+        # Replace by individuals from the best
+        lower_bound = all_fitnesses[-self._num_to_replace_from_best]
+        cond = fitnesses <= lower_bound
+
+        training_states = jax.tree_util.tree_map(
+            lambda x, y: jnp.where(
+                jnp.expand_dims(
+                    cond, axis=tuple([-(i + 1) for i in range(x.ndim - 1)])
+                ),
+                x,
+                y,
+            ),
+            best_training_states,
+            training_states,
+        )
+        replay_buffers = jax.tree_util.tree_map(
+            lambda x, y: jnp.where(
+                jnp.expand_dims(
+                    cond, axis=tuple([-(i + 1) for i in range(x.ndim - 1)])
+                ),
+                x,
+                y,
+            ),
+            best_replay_buffers,
+            replay_buffers,
+        )
+
+        # Replacing with samples from the ME repertoire
+        if self._num_to_replace_from_samples > 0:
+            me_samples, random_key = repertoire.sample(
+                random_key, self._config.pg_population_size_per_device
+            )
+            # Resample hyper-params
+            me_samples = jax.vmap(me_samples.__class__.resample_hyperparams)(me_samples)
+            upper_bound = all_fitnesses[
+                -self._num_to_replace_from_best - self._num_to_replace_from_samples
+            ]
+            cond = jnp.logical_and(fitnesses <= upper_bound, fitnesses >= lower_bound)
+            training_states = jax.tree_util.tree_map(
+                lambda x, y: jnp.where(
+                    jnp.expand_dims(
+                        cond, axis=tuple([-(i + 1) for i in range(x.ndim - 1)])
+                    ),
+                    x,
+                    y,
+                ),
+                me_samples,
+                training_states,
+            )
+
+        # Train the agents
+        env_states = emitter_state.env_states
+        # Init optimizers state before training the population
+        training_states = jax.vmap(training_states.__class__.init_optimizers_states)(
+            training_states
+        )
+        (training_states, env_states, replay_buffers), metrics = self._train_fn(
+            training_states, env_states, replay_buffers
+        )
+        # Empty optimizers states to avoid storing the info in the RAM
+        # and having too heavy repertoires
+        training_states = jax.vmap(training_states.__class__.empty_optimizers_states)(
```

Comment:
is this really useful? seems that this replace by zeros, zeros do not take space in the memory? 

Also, is it not a problem to change the shape of the arrays? (seems that the empty_optimizers_states implemented in pbt_td3 gives the shape (1,).

Reply:
yes it is 
you remove it and the code stops working with the hyper-parameters from the paper
the point is not that zeros take less space but rather that we store only single placeholders numbers instead of full arrays
as an example, typically the networks we use have around 300k parameters and an optimizer such as Adam stores 3x the network weights. With a repertoire of 1028 cells and for an actor-critic algorithm you may have to store 2 (actor + critic) x 3 (optimizer states) x 300k x 1028 float32 
here we save 2 x 3 x 1 x 1028 instead
changing the shape of an array is not an issue as long as all shapes remain fixed in advance
if the target shape would be variable then it would become an issue

Reply:
ok, makes sense!